solr/contrib/ltr/src/java/org/apache/solr/ltr/interleaving/TeamDraftInterleaving.java

```diff
@@ -0,0 +1,121 @@
+    // based on the current seed
+    String seed = System.getProperty("tests.seed");
+    if (seed == null) {
+      RANDOM = new Random();
+    } else {
+      RANDOM = new Random(seed.hashCode());
+    }
+  }
+
+  /**
+   * Team Draft Interleaving considers two ranking models: modelA and modelB.
+   * For a given query, each model returns its ranked list of documents La = (a1,a2,...) and Lb = (b1, b2, ...).
+   * The algorithm creates a unique ranked list I = (i1, i2, ...).
+   * This list is created by interleaving elements from the two lists la and lb as described by Chapelle et al.[1].
+   * Each element Ij is labelled TeamA if it is selected from La and TeamB if it is selected from Lb.
+   * <p>
+   * [1] O. Chapelle, T. Joachims, F. Radlinski, and Y. Yue.
+   * Large-scale validation and analysis of interleaved search evaluation. ACM TOIS, 30(1):1–41, Feb. (2012)
+   * <p>
+   * Assumptions:
+   * - rerankedA and rerankedB has the same length.
+   * They contains the same search results, ranked differently by two ranking models
+   * - each reranked list can not contain the same search result more than once.
+   *
+   * @param rerankedA a ranked list of search results produced by a ranking model A
+   * @param rerankedB a ranked list of search results produced by a ranking model B
+   * @return the interleaved ranking list
+   */
+  public InterleavingResult interleave(ScoreDoc[] rerankedA, ScoreDoc[] rerankedB) {
+    LinkedHashSet<ScoreDoc> interleavedResults = new LinkedHashSet<>();
+    ScoreDoc[] interleavedResultArray = new ScoreDoc[rerankedA.length];
+    ArrayList<Set<Integer>> interleavingPicks = new ArrayList<>(2);
+    Set<Integer> teamA = new HashSet<>();
+    Set<Integer> teamB = new HashSet<>();
+    int topN = rerankedA.length;
+    int indexA = 0, indexB = 0;
+
+    while (interleavedResults.size() < topN && indexA < rerankedA.length && indexB < rerankedB.length) {
+      if(teamA.size()<teamB.size() || (teamA.size()==teamB.size() && !RANDOM.nextBoolean())){
+        indexA = updateIndex(interleavedResults,indexA,rerankedA);
+        interleavedResults.add(rerankedA[indexA]);
+        teamA.add(rerankedA[indexA].doc);
+        indexA++;
+      } else{
+        indexB = updateIndex(interleavedResults,indexB,rerankedB);
+        interleavedResults.add(rerankedB[indexB]);
+        teamB.add(rerankedB[indexB].doc);
+        indexB++;
+      }
+    }
+    interleavingPicks.add(teamA);
+    interleavingPicks.add(teamB);
+    interleavedResultArray = interleavedResults.toArray(interleavedResultArray);
+
+    return new InterleavingResult(interleavedResultArray,interleavingPicks);
+  }
+
+  private int updateIndex(LinkedHashSet<ScoreDoc> interleaved, int index, ScoreDoc[] reranked) {
+    boolean foundElementToAdd = false;
+    while (index < reranked.length && !foundElementToAdd) {
+      ScoreDoc elementToCheck = reranked[index];
+      if (interleaved.contains(elementToCheck)) {
```

Comment:
Wondering re: `interleaved.contains(elementToCheck)` vs. `interleaved.contains(elementToCheck).doc` here. I note that the new `ScoreDoc.equals` method considers `doc` and `shardIndex` but not `score` which makes sense in the context here but for `ScoreDoc` alone is perhaps less obvious. The `interleavingPicks` use `doc` only (i.e. not `doc+shardIndex`). I haven't fully thought it through but intuitively if `doc+shardIndex` is required here then would it not also be required for the interleaving picks to avoid a strange edge case when potentially modelA picks "doc 1 from shard 1" and modelB picks "doc 1 from shard 2"?

Reply:
You are right, ScoreDoc.equals alone may be less obvious.
The reason I originally did that was to support sharding, but to be honest, currently Interleaving doesn't support sharding anyway (and adding the support is an extremely difficult topic that would require some serious thinking).

So I rolled back the ScoreDoc, and added an additional data structure on the doc Id, to do the contains.

Let me know if is better!

Reply:
> ... currently Interleaving doesn't support sharding ...

Let's include that in the documentation somehow, e.g. https://github.com/cpoerschke/lucene-solr/commit/7edd6005d6daa29b5a80bfd2a0e1249c23bd9279 say.

> ... So I rolled back the ScoreDoc ... Let me know if it is better!

Yes it is, thanks!